flow/beta/cli/config/config.go

```diff
@@ -83,33 +160,177 @@ func (k EmulatorServiceKey) MarshalJSON() ([]byte, error) {
 	})
 }
 
-type Network struct {
-	Host    string       `json:"host"`
-	ChainID flow.ChainID `json:"chain"`
+//UnmarshalJSON collection to our structure
+func (d *DeployCollection) UnmarshalJSON(b []byte) error {
+	raw := make(map[string]map[string]json.RawMessage)
+	d.Deploys = make([]Deploy, 0)
+
+	err := json.Unmarshal(b, &raw)
+	if err != nil {
+		return err
+	}
+
+	// go over each network
+	for network, v := range raw {
+		// for each network go through all accounts
+		for account, c := range v {
+			deploy := new(Deploy)
+			deploy.Network = network
+			deploy.Account = account
```

Comment:
It looks like you implemented these custom `UnmarshalJSON` functions to be able to assign the object keys (`network`, `account`) as values on the resulting `Deploy` struct. Is that correct?

I like the new `Config` struct you defined. The flatter structure is much easier to work with. However, I worry that the encoding/decoding logic is now a bit confusing -- it kind of parses and transforms the data at the same time.

I know you're working on refactoring the config logic already, so feel free to ignore this if you have a different approach, but I wanted to suggest that we define the JSON config structure with a separate `jsonConfig` type 👇 

(BTW, with this approach you don't need to define the `*Collection` structs.)

```go
type Config struct {
  Contracts Contracts
  // ...
}

type Contracts []Contract

type jsonConfig struct {
  Contracts jsonContracts `json:"contracts"`
  // ...
}

func (j jsonConfig) transformToConfig() Config {
  return Config{
    Contracts: j.Contracts.transformToConfig(),
    // ...
  }
}

type jsonContracts map[string]jsonContract

func (j jsonContracts) transformToContracts() Contracts {
  contracts := make(Contracts, 0)
  
  for contractName, c := range j {
    if c.Source != "" {
      contract := Contract{
        Name: contractName,
        Source: source,
      }
      
      contracts = append(contracts, contract) 
    }
    
    for networkName, source := range c.SourcesByNetwork {
      contract := Contract{
        Name: contractName,
        Source: source,
        Network: networkName,
      }
      
      contracts = append(contracts, contract) 
    }
    
    return contracts
}

type jsonContract struct {
    Source           string
    SourcesByNetwork map[string]string
}

func (j *jsonContract) UnmarshalJSON(b []byte) error {
  // ref: https://irshadhasmat.medium.com/golang-simple-json-parsing-using-empty-interface-and-without-struct-in-go-language-e56d0e69968
  var val interface{}
  
  err := json.Unmarshal(b, &val)
  if err != nil {
    return err
  }
  
  switch typedVal := val.(type) {
  case string:
    // simple: single source for all networks
    j.Source = typedVal
  case map[string]string:
    // advanced: different source for each network
    j.SourcesByNetwork = typedVal
  default:
    return errors.New("invalid contract definition")
  }

  return nil
}
```